### PR TITLE
nuv: init at 3.0.1-beta.2405292059

### PR DESCRIPTION
**Pull request by d4rkstar**
## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [X] aarch64-linux
  - [X] x86_64-darwin
  - [X] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [X] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: maintainers/maintainer-list.nix
Comment (AndersonTorres) [inline maintainers/maintainer-list.nix:45]: What are you doing here??

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix
Comment (d4rkstar) [reply maintainers/maintainer-list.nix:45]: Sorry, @AndersonTorres :( Don't know why that happened. Will fix ASAP

**Comment by d4rkstar** [comment]
Hello @AndersonTorres ! Is there any problem preventing the merge?

**Comment by AndersonTorres** [comment]
Put the attributes in the same order of this guide first.

https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations%2Fattribute-ordering.md

File: pkgs/by-name/nu/nuv/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/nu/nuv/package.nix:26]:
```suggestion
buildGoModule {
  pname = "nuv";
  inherit version;

```

File: pkgs/by-name/nu/nuv/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/nu/nuv/package.nix:32]:
```suggestion
    hash = "sha256-MdnBvlA4S2Mi/bcbE+O02x+wvlIrsK1Zc0dySz4FB/w=";
```

File: pkgs/by-name/nu/nuv/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/nu/nuv/package.nix:44]:
```suggestion
  # <explain why the check is set to false>
  doCheck = false;
```

File: pkgs/by-name/nu/nuv/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/nu/nuv/package.nix:59]: It is better to move these lines next to `src`.

```suggestion
```

File: pkgs/by-name/nu/nuv/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/nu/nuv/package.nix:34]: ... like this!

```suggestion

  vendorHash = "sha256-JkQbQ2NEaumXbAfsv0fNiQf/EwMs3SDLHvu7c/bU7fU=";
  subPackages = [ "." ];
  
```

File: pkgs/by-name/nu/nuv/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/nu/nuv/package.nix:61]:
```suggestion
  meta = {
```

File: pkgs/by-name/nu/nuv/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/nu/nuv/package.nix:66]: Quasi-ascii-betical.

```suggestion
    homepage = "https://nuvolaris.io/";
    description = "Nuvolaris Almighty CLI tool";
    license = lib.licenses.asl20;
    mainProgram = "nuv";
    maintainers = with lib.maintainers; [ msciabarra d4rkstar ];
```

File: pkgs/by-name/nu/nuv/tests.nix
Comment (AndersonTorres) [inline pkgs/by-name/nu/nuv/tests.nix:9]: can you use mktemp https://www.gnu.org/software/autogen/mktemp.html?

```suggestion
  export NUV_ROOT=$(mktemp -d /tmp/.nuv-XXXXX)
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/nu/nuv/package.nix
Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix]: Moved these lines as suggested. Let me know is ok now

File: pkgs/by-name/nu/nuv/package.nix
Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix:35]: these two lines are after src. I leaved them in alphabetical order. Let me know if it's ok

File: pkgs/by-name/nu/nuv/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/nu/nuv/package.nix:23]:
```suggestion
buildGoModule {
```

Is this rec needed?

File: pkgs/by-name/nu/nuv/tests.nix
Comment (AndersonTorres) [inline pkgs/by-name/nu/nuv/tests.nix:11]: Why is this creation of directories needed?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/nu/nuv/package.nix
Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix:23]: Hello, that "rec" is needed because the passthru tests inherit pname version;

File: pkgs/by-name/nu/nuv/tests.nix
Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/tests.nix:11]: Yes, these are needed because the "nuv" command attempts to download some tasks from another repository during startup. This is by design. To have the program version without network activity, is needed to emulate the presences of that directory.

File: pkgs/by-name/nu/nuv/package.nix
Comment (AndersonTorres) [reply pkgs/by-name/nu/nuv/package.nix:23]: What about putting them in the let-in already in use instead?

Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix]: Done. Let me know if now is ok!

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Finally, squash all commits:

1. One for each new contributor, with the message `maintainers:  add <maintainer>`
2. One for the package itself, `nuv: init at 3.0.1-beta.2405292059`


reports from review tool:

```json
{
    "blacklisted": [],
    "broken": [],
    "built": [
        "nuv"
    ],
    "extra-nixpkgs-config": null,
    "failed": [],
    "non-existent": [],
    "pr": 316731,
    "system": "x86_64-linux",
    "tests": []
}
```

**Comment by d4rkstar** [comment]
Sorry @AndersonTorres, I made a mistake with my branch and I'm sending the three commits back to a new PR

**Review by AndersonTorres — APPROVED** [review]

**Comment by d4rkstar** [comment]
Thank you @AndersonTorres !

**Comment by d4rkstar** [comment]
Hello @AndersonTorres there is something that prevent this PR from being merged?

**Comment by AndersonTorres** [comment]
Yes.  Namely, a committer :)

**Comment by h7x4** [comment]
Could @sciabarracom verify consent to being added to the maintainer list?

**Comment by sciabarracom** [comment]
Yes I confirm I want to be added

**Comment by d4rkstar** [comment]
Hello @h7x4, @sciabarracom confirmed. Let me know if something more is needed. Regards

File: pkgs/by-name/nu/nuv/package.nix
Comment (h7x4) [inline pkgs/by-name/nu/nuv/package.nix:54]: Do you think it would be better to wrap the `nuv` executable with a wrapper that sets `NUV_BIN` to a `symlinkJoin` of these packages? Or is there a good reason you've done it in this specific way?

File: pkgs/by-name/nu/nuv/package.nix
Comment (h7x4) [inline pkgs/by-name/nu/nuv/package.nix:63]: This doesn't really say anything unless you know what the tool is already. Could we write something like this instead?

```suggestion
    description = "A CLI tool for running tasks using the Nuvolaris serverless engine";
```

Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix:63]: Yes: that sounds better!

File: pkgs/by-name/nu/nuv/package.nix
Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix:54]: @h7x4 nuv command will do that by himself. So it's not necessary another wrapper

Comment (h7x4) [reply pkgs/by-name/nu/nuv/package.nix:54]: I'm not too worried about whether `nuv` picks it up or not. The thing that bothers me is the executable name clash. Let's say I was a user, and had constructed this environment:

```
pkgs.mkShell {
  packages = with pkgs; [
    nuv
    (kubectl.override {
      ... # misc overrides and patches to kubectl
    })
    ...
  ];
};
```

And inside of this environment, I run `kubectl`. I would expect my shell to run the overridden one, but AFAIK there is no good guarantee that the shell won't pick up the "old kubectl" from within `${nuv}/bin` first.

From what I'm reading in the upstream README, you can specify where the binary looks for executables with the `NUV_BIN` envvar, so I suggest we do something like this instead:

```nix
postInstall = let
  nuv-bin = pkgs.symlinkJoin {
    name = "nuv-bin";
    paths = [
      coreutils
      kubectl
      eksctl
      kind
      k3sup
    ];
  };
in ''
  makeWrapper $out/bin/nuv \
    --set NUV_BIN "${nuv-bin}/bin"
'';
```

This way, we can isolate `nuv`s executables, and not mess up the rest of the environment.

Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix:54]: @h7x4 thank you for insisting and for your observations: they are convincing and in the meantime I have studied the matter better. I'll test it with your changes and then we'll move on.

Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix:54]: Hello @h7x4 i tried your code inside the package.nix and it ended with an error:

```nix
Running phase: configurePhase
Running phase: buildPhase
Building subPackage ./.
Running phase: installPhase
dirname: unrecognized option '--set'
Try 'dirname --help' for more information.
mkdir: cannot create directory '': No such file or directory
/nix/store/1bmd2z777d232v0ikizk1h46vq881q0k-make-shell-wrapper-hook/nix-support/setup-hook: line 131: pop_var_context: head of shell_variables not a function context
build failed in installPhase with exit code 1
```

it seems that the makeWrapper command is missing something. Can you help me with that?

Comment (h7x4) [reply pkgs/by-name/nu/nuv/package.nix:54]: Sorry, `wrapProgram` seems more suitable. It's a version of `makeWrapper` that automatically renames the unwrapped executable inplace.

Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix]: @h7x4 check if my last commit is ok for you. I need still some test.

Comment (h7x4) [reply pkgs/by-name/nu/nuv/package.nix]: The code looks nice, the derivation builds. I have no idea how to use the tool though, so it would be nice if you could verify that it still runs as expected, and squash the changes into `nuv: init at 3.0.1-beta.2405292059`

Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix]: I agree: I will do some extra test and let you know.

Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix]: Hello @h7x4. I did my tests and here is the outcome.

I've builded a shell with the following `default.nix`: 

```nix
{ pkgs ? import <nixpkgs> {}, 
  version ? ""
}:
let  
  nuv = pkgs.callPackage ./nuv.nix {}; 
in
pkgs.mkShell {
  buildInputs = [
    nuv
    pkgs.which
    pkgs.curl
    pkgs.openssh
  ];
}
```
Curl and openssh are requirements for nuv to run.
So i've which and nuv commands. Started it with --pure.

```bash
bruno:tmp.BwyxQT05Mi/ $ nix-shell --pure default.nix                                                                  [9:19:41]

[nix-shell:/run/user/1000/tmp.BwyxQT05Mi]$ which nuv
/nix/store/8f22dwscgdsdji3sv1i5n7ap52mkwg9v-nuv-3.0.1-beta.2405062139/bin/nuv

[nix-shell:/run/user/1000/tmp.BwyxQT05Mi]$ strings /nix/store/8f22dwscgdsdji3sv1i5n7ap52mkwg9v-nuv-3.0.1-beta.2405062139/bin/nuv
#! /nix/store/306znyj77fv49kwnkpxmb0j2znqpa8bj-bash-5.2p26/bin/bash -e
export NUV_BIN='/nix/store/96rn7wbpa5gv85wlxbb3wz4assbznp8m-nuv-bin/bin'
exec -a "$0" "/nix/store/8f22dwscgdsdji3sv1i5n7ap52mkwg9v-nuv-3.0.1-beta.2405062139/bin/.nuv-wrapped"  "$@" 
```

The nuv command is wrapped and the NUV_BIN path have been set: all fine so far :)

Now if i list the content of the NUV_BIN folder there are all the symlinks we expect:

```bash
[nix-shell:/run/user/1000/tmp.BwyxQT05Mi]$ ls -lah /nix/store/96rn7wbpa5gv85wlxbb3wz4assbznp8m-nuv-bin/bin | grep nix
lrwxrwxrwx 1 root root   75 Jan  1  1970 .k3sup-wrapped -> /nix/store/c32c2fri7wgjv06iv5kydv42nl4ili41-k3sup-0.13.5/bin/.k3sup-wrapped
lrwxrwxrwx 1 root root   71 Jan  1  1970 coreutils -> /nix/store/php4qidg2bxzmm79vpri025bqi0fa889-coreutils-9.5/bin/coreutils
lrwxrwxrwx 1 root root   69 Jan  1  1970 eksctl -> /nix/store/y67k1mj8fh7i4x2ra0dl6i38jazqyi1d-eksctl-0.179.0/bin/eksctl
lrwxrwxrwx 1 root root   66 Jan  1  1970 k3sup -> /nix/store/c32c2fri7wgjv06iv5kydv42nl4ili41-k3sup-0.13.5/bin/k3sup
lrwxrwxrwx 1 root root   64 Jan  1  1970 kind -> /nix/store/8if8dwbvyalf0cfk0vkaysvcmkva9syh-kind-0.23.0/bin/kind
lrwxrwxrwx 1 root root   70 Jan  1  1970 kubectl -> /nix/store/05k95p72niq1gh39gv1g9n0ivsgl4hiy-kubectl-1.30.1/bin/kubectl
```

The nuv command seems to find kubectl and the other commands in it's NUV_BIN path:

```bash
[nix-shell:/run/user/1000/tmp.BwyxQT05Mi]$ nuv util kubectl help
kubectl controls the Kubernetes cluster manager.

 Find more information at: https://kubernetes.io/docs/reference/kubectl/
[...]
```

It seems to work correctly. I need to ask @sciabarracom if there are other implications with this before proceed with the solution.

Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix]: Hello @h7x4. It's ok for us to proceed. Let me know if you need to change something more.

Comment (h7x4) [reply pkgs/by-name/nu/nuv/package.nix]: Good to hear that it works, just squash the changes into `nuv: init at 3.0.1-beta.2405292059`, and I can merge it

**Comment by h7x4** [comment]
@d4rkstar are you familiar with the git operations you need to squash the commits? Feel free to ask if you're unsure

**Comment by d4rkstar** [comment]
> @d4rkstar are you familiar with the git operations you need to squash the commits? Feel free to ask if you're unsure

Yes, and it's usually a trivial task but for some reason  the squash ends up with a merge from master. :( I'm doing something the wrong way

**Comment by h7x4** [comment]
Can you do `git rebase -i 5d09f92c53c8bbc5fc14c8d673b233c217ea05a9` (the hash of the commit above the first `nuv: init at 3.0.1-beta.2405292059`) and then mark all commits except that one as `f` or `fixup`? That should do it.

In the meanwhile, I'll temporarily mark the PR as draft in case something goes wrong, so we avoid mass pinging people.

**Review by h7x4 — APPROVED** [review]

**Review by h7x4 — COMMENTED** [review]
Just a few things I missed before clicking the approve button last time

File: pkgs/by-name/nu/nuv/package.nix
Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix]: Yes, this is a good suggestion. Thank you

File: pkgs/by-name/nu/nuv/package.nix
Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/package.nix]: done. Let me know if it's ok for you

File: pkgs/by-name/nu/nuv/tests.nix
Comment (d4rkstar) [reply pkgs/by-name/nu/nuv/tests.nix]: also this cleanup is done!

**Review by h7x4 — APPROVED** [review]

**Pull request merged**
